### ISSUE TITLE
Generalised types can be ill-formed

### ISSUE THREAD

**Issue by nad** [agda/agda#3667]
Consider the following code:
```agda
data Unit : Set where
  unit : Unit

P : Unit → Set
P unit = Unit

variable
  u : Unit
  p : P u

postulate
  Q : (u : Unit) → P u → Set
  q : P u → Q u p
```
The inferred type of `q`:
```
{u p.u : Unit} {p : P p.u} → P u → Q u p
```
Note that this type is ill-formed.

**Comment by UlfNorell** [comment]
The problem is that generalization doesn't handle constraints properly.

The type that it gets to work with is this:
```agda
postulate
  q : let u = _
          p : ∀ {u} → P u
          p = _
      in P u → Q u p
```
which has the unsolved constraints
```
Failed to solve the following constraints:
  _21 := _17 [blocked on problem 20]
  [20] P _u_20 =< P _13 : Set
```
At this point two things go wrong:
- we decide to generalize over both `_20` and `_13` (causing #3666), and
- we lose the constraint (causing this issue)

**Comment by jespercockx** [comment]
It seems dangerous/wrong to generalize over constrained metavariables in the first place. Why don't we just postpone generalization until there are no more constraints left (or show the unsolved constraints to the user in case they're never solved)?

**Comment by UlfNorell** [comment]
The constraints are local, so postponing doesn't help. Also it's problematic to postpone generalization: how would you check the lhs of the function if you don't have a type?

The right thing is indeed to not generalize over constrained metas.

**Comment by UlfNorell** [comment]
@nad I have a fix that solves the problem here and in #3666. Can you try #3670 on your original example (I assume some shrinking happened to obtain the test case in the OP)?

**Comment by nad** [comment]
Some code that used to be accepted is now rejected. I tried to construct a small test case (but it might not be entirely representative of the original code):
```agda
data D (A : Set) : Set where
  d : D A

postulate
  E : Set

P : (A : Set) → D A → Set
P _ d = E

postulate
  Q : (A : Set) (x : D A) → P A x → Set

variable
  A : Set
  x : D A
  p : P A x

postulate
  q : Q _ x p
```
There is one unsolved metavariable, `_x.A_25`, and no unsolved constraints. The inferred type of `q` is `{x : D _x.A_25} {p : P _x.A_25 x} → Q _x.A_25 x p`.

**Comment by nad** [comment]
I retract the previous comment, I seem to have used the wrong version of Agda when testing.

**Comment by nad** [comment]
I retract the retraction. (Sorry for the confusion.)

**Comment by UlfNorell** [comment]
This is an expected side effect of the fix. At the point of generalisation there's a constraint on the first argument to `Q` preventing it from being generalised. Once we generalise over `x` injectivity kicks in and the constraint is solved, which is why you don't see it.

This could be fixed by treating metas from generalisable variables as neutral, can you file a separate enhancement ticket for that?

**Comment by andreasabel** [comment]
See #3694 where the handling of constraints is still not correct.
